src/System.Private.ServiceModel/tools/test/SelfHostWcfService/TestResources/DigestServiceAuthorizationManager.cs

```diff
@@ -0,0 +1,307 @@
+            }
+
+            string password;
+            if (!GetPassword(digestState.Username, out password))
+            {
+                return UnauthorizedResponse(digestState);
+            }
+
+            digestState.Password = password;
+            if (!digestState.Authorized || digestState.StaleNonce)
+            {
+                return UnauthorizedResponse(digestState);
+            }
+
+            return Authorized(digestState, operationContext, ref message);
+        }
+
+        protected virtual DateTime GetNonceExpiryTime()
+        {
+            return DateTime.Now + _nonceValidityTime;
+        }
+
+        public abstract bool GetPassword(string username, out string password);
+
+        private bool UnauthorizedResponse(DigestAuthenticationState digestState)
+        {
+            digestState.NonceExpiryTime = GetNonceExpiryTime();
+            digestState.SetChallengeResponse(HttpStatusCode.Unauthorized, "Access Denied");
+            return false;
+        }
+
+        #region IServiceBehavior
+        public void Validate(ServiceDescription serviceDescription, ServiceHostBase serviceHostBase)
+        {
+            return;
+        }
+
+        public void AddBindingParameters(ServiceDescription serviceDescription, ServiceHostBase serviceHostBase, Collection<ServiceEndpoint> endpoints,
+            BindingParameterCollection bindingParameters)
+        {
+            return;
+        }
+
+        public void ApplyDispatchBehavior(ServiceDescription serviceDescription, ServiceHostBase serviceHostBase)
+        {
+            serviceHostBase.Authorization.ServiceAuthorizationManager = this;
+            foreach (ChannelDispatcherBase t in serviceHostBase.ChannelDispatchers)
+            {
+                ChannelDispatcher channelDispatcher = t as ChannelDispatcher;
+                foreach (EndpointDispatcher endpointDispatcher in channelDispatcher.Endpoints)
+                {
+                    endpointDispatcher.DispatchRuntime.ServiceAuthorizationManager = this;
+                }
+            }
+        }
+        #endregion
+
+        private struct DigestAuthenticationState
+        {
+            private string _authorizationHeader;
+            private OperationContext _operationContext;
+            private Dictionary<string, string> _authorizationParameters;
+            private string _method;
+            private string _nonceString;
+            private string _password;
+            private string _realm;
+            private bool? _authorized;
+
+            public DigestAuthenticationState(OperationContext operationContext, string realm)
+            {
+                _operationContext = operationContext;
+                _realm = realm;
+                _password = null;
+                _authorized = new bool?();
+                _authorizationHeader = GetAuthorizationHeader(operationContext, out _method);
+                if (_authorizationHeader.Length < 7)
+                {
```

Comment:
It's 7 because the word "Digest " with a trailing space is 7 characters long :). I'll add a constant.